src/openfermion/utils/_davidson.py

```diff
@@ -0,0 +1,273 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""This module is to find lowest eigenvalues with Davidson algorithm."""
+from __future__ import absolute_import
+
+import numpy
+import numpy.linalg
+import scipy
+import scipy.linalg
+import scipy.sparse.linalg
+
+from openfermion.utils._sparse_tools import (get_linear_qubit_operator,
+                                             get_linear_qubit_operator_diagonal)
+
+# Exceptions.
+class DavidsonError(Exception):
+    """Exceptions."""
+    pass
+
+
+class Davidson(object):
+    """Davidson algorithm to get the n states with smallest energies."""
+
+    def __init__(self, linear_operator, linear_operator_diagonal, eps=1e-6):
+        """
+        Args:
+            linear_operator(scipy.sparse.linalg.LinearOperator): The linear
+                operator which defines a dot function when applying on a vector.
+            linear_operator_diagonal(numpy.ndarray): The linear operator's
+                diagonal elements.
+            eps(float): The max error for eigen vectors' elements during
+                iterations.
+        """
+        if not isinstance(linear_operator, scipy.sparse.linalg.LinearOperator):
+            raise ValueError(
+                'linear_operator is not a LinearOperator: {}.'.format(type(
+                    linear_operator)))
+
+        self.linear_operator = linear_operator
+        self.linear_operator_diagonal = linear_operator_diagonal
+        self.eps = eps
+
+    def get_lowest_n(self, n_lowest=1, initial_guess=None, max_iterations=300):
+        """
+        Returns `n` smallest eigenvalues and corresponding eigenvectors for
+            linear_operator.
+
+        Args:
+            n(int): The number of states corresponding to the smallest eigenvalues
+                and associated eigenvectors for the linear_operator.
+            initial_guess(numpy.ndarray[complex]): Initial guess of eigenvectors
```

Comment:
If I'm guessing right, this should have n columns, each of which is a guess at an eigenvector.  Do we need to put any constraints on this?  For example, do they need to be orthogonal for success?

We should maybe also check if initial_guess.shape[1] < n but not none, and pass a warning that we need to generate random vectors for the rest.

Reply:
1. The only constraints is that all guess_v vectors are not zero. Added check.
2. Added random vectors in this case, though now really required.